datafusion-examples/examples/advanced_udaf.rs

```diff
@@ -0,0 +1,228 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use datafusion::{arrow::datatypes::DataType, logical_expr::Volatility};
+use std::{any::Any, sync::Arc};
+
+use arrow::{
+    array::{ArrayRef, Float32Array},
+    record_batch::RecordBatch,
+};
+use datafusion::error::Result;
+use datafusion::prelude::*;
+use datafusion_common::{cast::as_float64_array, ScalarValue};
+use datafusion_expr::{Accumulator, AggregateUDF, AggregateUDFImpl, Signature};
+
+/// This example shows how to use the full AggregateUDFImpl API to implement a user
+/// defined aggregate function. As in the `simple_udaf.rs` example, this struct implements
+/// a function `accumulator` that returns the `Accumulator` instance.
+///
+/// To do so, we must implement the `AggregateUDFImpl` trait.
+#[derive(Debug, Clone)]
+struct GeoMeanUdf {
+    signature: Signature,
+}
+
+impl GeoMeanUdf {
+    /// Create a new instance of the GeoMeanUdf struct
+    fn new() -> Self {
+        Self {
+            signature: Signature::exact(
+                // this function will always take one arguments of type f64
+                vec![DataType::Float64],
+                // this function is deterministic and will always return the same
+                // result for the same input
+                Volatility::Immutable,
+            ),
+        }
+    }
+}
+
+impl AggregateUDFImpl for GeoMeanUdf {
+    /// We implement as_any so that we can downcast the AggregateUDFImpl trait object
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    /// Return the name of this function
+    fn name(&self) -> &str {
+        "geo_mean"
+    }
+
+    /// Return the "signature" of this function -- namely that types of arguments it will take
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    /// What is the type of value that will be returned by this function.
+    fn return_type(&self, _arg_types: &[DataType]) -> Result<DataType> {
+        Ok(DataType::Float64)
+    }
+
+    /// This is the accumulator factory; DataFusion uses it to create new accumulators.
+    fn accumulator(&self, _arg: &DataType) -> Result<Box<dyn Accumulator>> {
```

Comment:
👍  

While this "advanced" usage isn't much more advanced than the current "simple" UDAF I think this PR now provides a home / plausible way to implement the full `GroupsAccumulator` API for UDAFs (which is the powerful, very performant API used by built in aggregate functions in DataFusion)

Reply:
Filed https://github.com/apache/arrow-datafusion/issues/8793 to track